crates/oxc_linter/src/rules/typescript/consistent_indexed_object_style.rs

```diff
@@ -0,0 +1,583 @@
+use oxc_ast::{
+    ast::{Statement, TSSignature, TSType, TSTypeName},
+    AstKind,
+};
+use oxc_diagnostics::{
+    miette::{self, Diagnostic},
+    thiserror::Error,
+};
+use oxc_macros::declare_oxc_lint;
+use oxc_span::Span;
+
+use crate::{context::LintContext, rule::Rule, AstNode};
+
+#[derive(Debug, Error, Diagnostic)]
+#[error("typescript-eslint(consistent-indexed-object-style):")]
+#[diagnostic(severity(warning), help("A {0} is preferred over an {1}."))]
+struct ConsistentIndexedObjectStyleDiagnostic(
+    &'static str,
+    &'static str,
+    #[label("A {0} is preferred over an {1}.")] pub Span,
+);
+
+#[derive(Debug, Default, Clone)]
+pub struct ConsistentIndexedObjectStyle {
+    is_index_signature: bool,
+}
+
+#[derive(Debug, Default, Clone, Copy, Eq, PartialEq)]
+enum ConsistentIndexedObjectStyleConfig {
+    #[default]
+    Record,
+    IndexSignature,
+}
+
+declare_oxc_lint!(
+    /// ### What it does
+    /// Require or disallow the `Record` type.
+    ///
+    /// ### Why is this bad?
+    ///
+    ///
+    /// ### Example
+    /// ```javascript
+    /// ```
+    ConsistentIndexedObjectStyle,
+    style
+);
+
+impl Rule for ConsistentIndexedObjectStyle {
+    fn from_configuration(value: serde_json::Value) -> Self {
+        let config = value.get(0).and_then(serde_json::Value::as_str).map_or_else(
+            ConsistentIndexedObjectStyleConfig::default,
+            |value| match value {
+                "record" => ConsistentIndexedObjectStyleConfig::Record,
+                _ => ConsistentIndexedObjectStyleConfig::IndexSignature,
+            },
+        );
+        Self { is_index_signature: config == ConsistentIndexedObjectStyleConfig::IndexSignature }
+    }
+
+    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
+        match node.kind() {
+            AstKind::TSInterfaceDeclaration(decl) => {
+                if decl.body.body.len() == 1 {
+                    for signature in &decl.body.body {
+                        if let TSSignature::TSIndexSignature(idx) = signature {
+                            match &idx.type_annotation.type_annotation {
```

Comment:
Why we need to check `idx.type_annotation.type_annotation`, look like typescript-ESLint only check [`parameter`'s type](https://github.com/typescript-eslint/typescript-eslint/blob/584db29ec44ce4e9cb71afac35d48994889168e6/packages/eslint-plugin/src/rules/consistent-indexed-object-style.ts#L48-L61).

Reply:
for the following case:
```ts
interface Foo {
   [key: string]: Foo;
}
```

even with a setting of `Record`, the above is still valid

Reply:
not sure I got what you meant here

Reply:
For case

```ts
interface Foo {
   [key: string]: Foo; // a
}
```

It is valid because it been treated as a circle, `Foo` in the second line refers to the interface name `Foo`.